subprojects/core/src/main/java/org/gradle/internal/vfs/AbstractEventDrivenFileWatcherRegistry.java

```diff
@@ -19,30 +19,48 @@
 import net.rubygrapefruit.platform.file.FileWatcher;
 import net.rubygrapefruit.platform.file.FileWatcherCallback;
 import org.gradle.internal.vfs.watch.FileWatcherRegistry;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import java.io.IOException;
 import java.nio.file.Path;
 import java.nio.file.Paths;
 import java.util.Set;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
 
 import static org.gradle.internal.vfs.watch.FileWatcherRegistry.Type.CREATED;
 import static org.gradle.internal.vfs.watch.FileWatcherRegistry.Type.INVALIDATE;
 import static org.gradle.internal.vfs.watch.FileWatcherRegistry.Type.MODIFIED;
 import static org.gradle.internal.vfs.watch.FileWatcherRegistry.Type.REMOVED;
 
 public class AbstractEventDrivenFileWatcherRegistry implements FileWatcherRegistry {
+    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractEventDrivenFileWatcherRegistry.class);
+
     private final FileWatcher watcher;
     private final AtomicInteger numberOfReceivedEvents = new AtomicInteger();
     private final AtomicBoolean unknownEventEncountered = new AtomicBoolean();
+    private final AtomicReference<Throwable> errorWhileReceivingFileChanges = new AtomicReference<>();
 
     public AbstractEventDrivenFileWatcherRegistry(Set<Path> roots, FileWatcherCreator watcherCreator, ChangeHandler handler) {
         this.watcher = createWatcher(roots, watcherCreator, handler);
     }
 
     private FileWatcher createWatcher(Set<Path> roots, FileWatcherCreator watcherCreator, ChangeHandler handler) {
-        FileWatcher watcher = watcherCreator.createWatcher((type, path) -> handleEvent(type, path, handler));
+        FileWatcher watcher = watcherCreator.createWatcher(new FileWatcherCallback() {
+            @Override
+            public void pathChanged(Type type, String path) {
+                handleEvent(type, path, handler);
+            }
+
+            @Override
+            public void reportError(Throwable ex) {
+                LOGGER.error("Error while receiving file changes", ex);
+                errorWhileReceivingFileChanges.compareAndSet(null, ex);
```

Comment:
Shall we collect all such errors in a list instead?

Reply:
We could, though I am not sure it is much use. We will log the errors on the native side as well, right?

I think the first error is enough.

Reply:
Alright.